### PR TITLE
Fix: Constant initialization in the Dafny-to-Rust code generator

### LINKED ISSUE #5833: constant assignment fails in Rust
### Dafny version

latest-nightly

### Code to produce this issue

```dafny
class X {
        const c: int
        constructor(c_init: int) {
            this.c := c_init;
        }
    }
```

### Command to run and resulting output

```code
dafny run -t:rs
```

### What happened?

This is a regression as it used to work

### What type of operating system are you experiencing the problem on?

Windows

### PR DESCRIPTION
**Pull request by MikaelMayer**
Fixes #5833
### Description
There are now three kinds of generated fields:
* Internal constant fields (internal const): They have the same type and assigned and read using regular field read and field update.
* Mutable fields (var): They are wrapped with `::dafny_runtime::Field<...>` to avoid any soundness issue, and use the wrappers `read_field` and `modify_field` to read and modify them.
* Constant fields (const): They are called with a function call with zero argument.

Before this PR, only the last two were implemented, which lead to issues when trying to assign the default value of a class constant.
This issue was not captured because the test started to fail but since it's was marked `%testForEachCompiler` and these tests are ignored in the CI, we did not detect it.

### How has this been tested?
Updated the file classes.dfy. All other tests in comp/rust should pass.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: docs/dev/news/5833.fix
Comment (robin-aws) [inline docs/dev/news/5833.fix:1]: Intentional?

Have we been announcing bug fixes like this while the Rust code generator isn't fully supported yet?

File: Source/DafnyCore/Backends/Dafny/AST.dfy [hunk 1]
```diff
@@ -323,6 +323,7 @@ module {:extern "DAST"} DAST {
     UnboundedIntRange(start: Expression, up: bool) |
     Quantifier(elemType: Type, collection: Expression, is_forall: bool, lambda: Expression)
 
+  datatype FieldMutability = ConstantField | MutableField | InternalConstantField
```
Comment (robin-aws) [inline Source/DafnyCore/Backends/Dafny/AST.dfy:326]:
```suggestion
  datatype FieldMutability = ConstantField | ClassMutableField | ClassConstantField
```

I didn't understand "InternalConstantField" until I looked at the test.

**Review by robin-aws — CHANGES_REQUESTED** [review]
Thanks for the quick fix. As a follow up can we re-enable the Rust backend in `%testDafnyForEachCompiler`?

File: docs/dev/news/5833.fix
Comment (MikaelMayer) [reply docs/dev/news/5833.fix:1]: I'd prefer to leave it quiet too.

File: Source/DafnyCore/Backends/Dafny/AST.dfy (hunk 1 above)
Comment (MikaelMayer) [reply Source/DafnyCore/Backends/Dafny/AST.dfy:326]: ClassMutableField is ok, but the last one is really internal, so it should be ClassInternalConstantField, all right?

File: Source/DafnyCore/Backends/Rust/Dafny-compiler-rust.dfy [hunk 2]
```diff
@@ -3298,11 +3298,14 @@ module {:extern "DCOMP"} DafnyToRustCompiler {
...
             r := r.Sel(escapeVar(field));
-            if isConstant {
+            if fieldMutability.ConstantField? {
               r := r.Apply0();
               r := r.Clone(); // self could be &mut, so to avoid any borrow checker problem, we clone the value.
+            } else if fieldMutability.InternalConstantField? {
+              r := r.Clone();
             } else {
-              r := read_field_macro.Apply1(r); // Already contains a clone.
+              assert fieldMutability.MutableField?;
```
Comment (robin-aws) [inline Source/DafnyCore/Backends/Rust/Dafny-compiler-rust.dfy:3307]: Wouldn't a match statement be a little cleaner?

File: Source/DafnyCore/Backends/Dafny/AST.dfy (hunk 1 above)
Comment (robin-aws) [reply Source/DafnyCore/Backends/Dafny/AST.dfy]: Thanks for adding the comment, that helps

**Review by robin-aws — APPROVED** [review]

**Pull request merged**